design/one-pager-resource-connectivity-mvp.md

```diff
@@ -0,0 +1,653 @@
+   ensure any managed resources backing the aforementioned claims can connect,
+   for example ensuring they're both configured to use the aforementioned VPC
+   network.
+1. The app operator creates their `KubernetesCluster` and `MySQLInstance`
+   resource claims, which either explicitly reference or default to the resource
+   classes created in step 2. Because their underlying managed resources are
+   appropriately configured, they can now communicate!
+
+The remainder of this proposal works through this scenario in each cloud
+provider, highlighting the changes necessary to enable connectivity.
+
+### Resource References
+
+It's typical for Kubernetes resources to refer to each other. The [Kubernetes
+API conventions] state:
+
+> References to loosely coupled sets of objects, such as pods overseen by a
+> replication controller, are usually best referred to using a label selector.
+> [...] References to specific objects, especially specific resource versions
+> and/or specific fields of those objects, are specified using the
+> ObjectReference type (or other types representing strict subsets of it).
+> [...] Object references should either be called fooName if referring to an
+> object of kind Foo by just the name (within the current namespace, if a
+> namespaced resource), or should be called fooRef, and should contain a subset
+> of the fields of the ObjectReference type.
+
+Crossplane currently uses these conventions to create references between
+resource claims and classes, resource claims and managed resources, etc. No
+pattern currently exists for modelling relationships between managed resources.
+
+Assume a `Network` managed resource named `kubernetesname` exists in namespace
+`crossplane-system`. Further assume this managed resource represents a GCP VPC
+network whose 'real' name in the GCP API is `externalname`. Finally, assume a
+`CloudSQLInstance` managed resource wants to specify this network as its
+`.spec.ipConfiguration.privateNetwork` (i.e. whitelisted VPC network). This
+could be implemented in one of two ways:
+
+1. The `CloudSQLInstance` simply requires `.spec.ipConfiguration.privateNetwork`
+   be provided as a `string` name specifying the real, external name of the
+   `Network`, i.e. `externalname`. It is up to the creator of the
+   `CloudSQLInstance` (or the `ResourceClass` used to create it) to ensure this
+   name corresponds with the name of the `kubernetesname` `Network`.
+1. The `CloudSQLInstance` requires an `ObjectReference` at be specified as
+   `.spec.ipConfiguration.privateNetworkRef`. This `ObjectReference` refers to
+   `kind: Network`, `namespace: crossplane-system`, `name: kubernetesname`. The
+   `CloudSQLInstance` controller must then lookup the specified `Network` in
+   order to determine that it is named `externalname` when submitting requests
+   to the GCP API.
+
+This document suggests the first approach, both because it requires less logic
```

Comment:
With the first approach, how would I connect a consumer in an Amazon region to a Google-managed CloudSQL instance? Does it come down to the configuration of the network in which the CloudSQL instance is running?

Reply:
I think this is still an open question that is not necessarily addressed in this proposal. Cross-provider connectivity is something that could be pretty tricky. I think it is important to be asking these kinds of questions now though because the decisions that we make on individual provider networking could constrain us down the line. My current thinking (has not be corroborated by anyone else) on cross-provider networks is likely that we will want to provide some layer of "meta networking" components over the top of provider specific ones. That being said, I think that individual provider networking components in Crossplane should be as high fidelity as possible, and the meta level should be designed to talk to each provider in its "language".

Reply:
@suskin, @hasheddan captured my thoughts here pretty well. I think it's important for Crossplane to support some base layer that is not clever or opinionated, but rather just does explicitly what you configure it to do. This ensures that:

* We have a base layer upon which to build clever logic that automatically configures network connectivity.
* When our clever logic automatically configures network connectivity, you'll be able to inspect the managed resources Crossplane created to determine what it did at the base layer.
* If someone doesn't like how our clever logic works, they can still use Crossplane to manually configure said base layer.
* If someone _really_ doesn't like how our clever logic works, they can build their own clever logic as a Kubernetes controller that manipulates our base layer.

I imagine we'll progress in building this somewhat like:

1. A cloud administrator can connect a Kubernetes cluster to a SQL database running in the same cloud provider and region, without ever leaving Crossplane, by configuring them correctly. (The scope of this MVP design.)
1. A cloud administrator can connect an arbitrary managed resource to another arbitrary managed resource running in the same cloud provider and region, without ever leaving Crossplane, by configuring them correctly.
1. A cloud administrator can connect an arbitrary managed resource to another arbitrary managed resource running in different cloud providers, without ever leaving Crossplane, by configuring them correctly.
1. An application operator can create two arbitrary resource claims and declare that they must communicate. Crossplane uses policy configured by a cloud administrator to automatically configure the managed resources underlying said claims (and any required interstitial managed resources), even if they're running in different clouds, to ensure they can communicate.

We started out trying to approach this from step 4, but it's such a huge problem space that we scoped it down to step 1 with a plan to take what we learn here and build on it.